### PR TITLE
Produce action fixes

### PR DESCRIPTION
**Pull request by bjustin-ibm**
- limited retry behavior when the producer fails to connect
- tweak producer config to be more efficient
- additional logging
- cache producer for reuse in certain circumstances

### CODE REVIEW

**Comment by bjustin-ibm** [comment]
@dubeejw please review

**Comment by bjustin-ibm** [comment]
Resolves: #182

**Comment by bjustin-ibm** [comment]
The retry path does not kick in if the initial connection to Kafka takes too long. Perhaps there is a timeout configuration that should be set lower?

**Comment by bjustin-ibm** [comment]
Reusing the cached producer requires a change in the Python 3 action runtime found here: https://github.com/apache/incubator-openwhisk/pull/2522

**Comment by bjustin-ibm** [comment]
@dnwe please review.

File: action/kafkaProduce.py [hunk 1]
```diff
@@ -18,46 +18,96 @@
...
+            print("Finding topic {}".format(topic))
+            partition_info = producer.partitions_for(topic)
+            print("Found topic {} with partition(s) {}".format(topic, partition_info))
+
+            break
+        except Exception as e:
+            if attempt == max_attempts:
+                producer = None
+                logging.warning(e)
+                traceback.print_stack(limit=5)
```
Comment (dnwe) [inline action/kafkaProduce.py]: You'll want `traceback.print_exc(limit=5)` here, to print the stack trace of the Exception that was raised. `print_stack` just dumps a stacktrace of where you currently are in your code, and won't show where the caught Exception came from

File: action/kafkaProduce.py [hunk 2]
```diff
@@ -18,46 +18,96 @@
...
+                future = producer.send(topic, bytes(value, 'utf-8'))
+
+            sent = future.get(timeout=20)
+            msg = "Successfully sent message to {}:{} at offset {}".format(
+                sent.topic, sent.partition, sent.offset)
+            print(msg)
+            result = {"success": True, "message": msg}
+        except Exception as e:
+            logging.warning(e)
+            traceback.print_stack(limit=5)
```
Comment (dnwe) [inline action/kafkaProduce.py]: As above, you want `print_exc` here I think.

File: action/messageHubProduce.py [hunk 3]
```diff
@@ -18,64 +18,97 @@
...
+                future = producer.send(topic, bytes(value, 'utf-8'))
+
+            sent = future.get(timeout=20)
+            msg = "Successfully sent message to {}:{} at offset {}".format(
+                sent.topic, sent.partition, sent.offset)
+            print(msg)
+            result = {"success": True, "message": msg}
+        except Exception as e:
+            logging.warning(e)
+            traceback.print_stack(limit=5)
```
Comment (dnwe) [inline action/messageHubProduce.py]: Same as the earlier comment about print_exc

File: action/messageHubProduce.py [hunk 4]
```diff
@@ -108,3 +147,64 @@ def validateParams(params):
...
+
+def getConnectionHash(params):
+    # always use the sorted brokers to combat the effects of shuffle()
+    brokers = params['kafka_brokers_sasl']
+    brokers.sort()
+    brokersString = ",".join(brokers)
+
+    apiKey = "{}:{}".format(params['user'], params['password'])
+
+    connectionHash = brokersString + apiKey
```
Comment (dnwe) [inline action/messageHubProduce.py]: Technically the broker URLs are only used for bootstrapping to fetch the cluster info. The client discoverers all the available brokers in the response to that bootstrap, so your connectionHash can just be based on the apiKey as that will only be valid for a given cluster

**Comment by bjustin-ibm** [comment]
@dnwe I addressed all of your review comments, thanks a bunch!
@dubeejw please review.

File: action/messageHubProduce.py [hunk 5]
```diff
@@ -196,7 +205,16 @@ def getProducer(validatedParams):
...
 
-
 def getConnectionHash(params):
     apiKey = "{}:{}".format(params['user'], params['password'])
     return apiKey
+
+# return the remaining time (in seconds) until the action will expire,
+# optionally reserving some time (also in seconds).
+def getRemainingTime(reservedTime=0):
+    deadlineSeconds = int(os.getenv('__OW_DEADLINE', 60000)) / 1000
```
Comment (dubeejw) [inline action/messageHubProduce.py]: `__OW_DEADLINE` is always a positive number?

Comment (bjustin-ibm) [inline action/messageHubProduce.py]: Yes, it is epoch milliseconds.

**Comment by bjustin-ibm** [comment]
@csantanapr please merge.

**Pull request merged**